File: src/commands/mcp.ts
import { createRequire } from "module";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";
import { insertPrompt, insertPromptsBatch, PromptRecord } from "../lib/supabase.js";
import { loadAuth } from "../lib/auth.js";

const require = createRequire(import.meta.url);
const pkg = require("../../package.json") as { version: string };

let captureCount = 0;

export async function runMcp(): Promise<void> {
  const auth = loadAuth();
  const userId = auth?.userId;

  const server = new McpServer({
    name: "pcr-dev",
    version: pkg.version,
  });

  server.registerTool(
    "pcr_log_prompt",
    {
      title: "Log Prompt",
      description:
        "Log a prompt and its AI response for later code review. Call this after completing a coding task.",
      inputSchema: {
        prompt_text: z.string().describe("The user's prompt or instruction to the AI"),
        response_text: z.string().optional().describe("The AI's response (summary or full text)"),
        session_id: z.string().optional().describe("Session identifier to group related prompts"),
        project_name: z.string().optional().describe("Name of the project being worked on"),
        branch_name: z.string().optional().describe("Git branch name"),
        model: z.string().optional().describe("AI model used (e.g., claude-sonnet-4-5)"),
        source: z.string().optional().describe("AI tool name (e.g., claude-code, cursor, codex)"),
        files_changed: z.array(z.string()).optional().describe("List of files modified"),
      },
    },
    async (params) => {
      const record: PromptRecord = {
        session_id: params.session_id || `mcp-${Date.now()}`,
        project_name: params.project_name,
        branch_name: params.branch_name,
        prompt_text: params.prompt_text,
        response_text: params.response_text,
        model: params.model,
        source: params.source || "unknown",
        capture_method: "mcp",
        file_context: params.files_changed ? { files: params.files_changed } : undefined,
        captured_at: new Date().toISOString(),
        user_id: userId,
      };

      const success = await insertPrompt(record);
      if (success) captureCount++;

      return {
        content: [
          {
            type: "text" as const,
            text: success
              ? `PCR: Prompt logged (session total: ${captureCount})`
              : "PCR: Failed to log prompt — check connection",
          },
        ],
      };
    }
  );

  server.registerTool(
    "pcr_log_session",
    {
      title: "Log Session",
      description:
        "Log an entire coding session transcript. Use at the end of a session to capture all interactions at once.",
      inputSchema: {
        session_id: z.string().describe("Unique session identifier"),
        project_name: z.string().optional().describe("Project name"),
        branch_name: z.string().optional().describe("Git branch"),
        source: z.string().optional().describe("AI tool name"),
        messages: z
          .array(
            z.object({
              role: z.enum(["user", "assistant"]),
              content: z.string(),
              model: z.string().optional(),
              timestamp: z.string().optional(),
            })
          )
          .describe("Array of messages in the session"),
      },
    },
    async (params) => {
      const records: PromptRecord[] = [];

      for (let i = 0; i < params.messages.length; i++) {
        const msg = params.messages[i];
        if (msg.role !== "user") continue;

        let responseText: string | undefined;
        let model: string | undefined;
        if (i + 1 < params.messages.length && params.messages[i + 1].role === "assistant") {
          responseText = params.messages[i + 1].content;
          model = params.messages[i + 1].model;
        }

        records.push({
          session_id: params.session_id,
          project_name: params.project_name,
          branch_name: params.branch_name,
          prompt_text: msg.content,
          response_text: responseText,
          model: model || msg.model,
          source: params.source || "unknown",
          capture_method: "mcp",
          captured_at: msg.timestamp || new Date().toISOString(),
          user_id: userId,
        });
      }

      const count = await insertPromptsBatch(records);
      captureCount += count;

      return {
        content: [
          {
            type: "text" as const,
            text: `PCR: Session logged — ${count} prompt(s) captured (session total: ${captureCount})`,
          },
        ],
      };
    }
  );

  server.registerTool(
    "pcr_status",
    {
      title: "PCR Status",
      description: "Check how many prompts have been captured in this PCR session",
    },
    async () => {
      return {
        content: [
          {
            type: "text" as const,
            text: `PCR.dev status: ${captureCount} prompt(s) captured this session. User: ${userId ?? "not logged in"}`,
          },
        ],
      };
    }
  );

  const transport = new StdioServerTransport();
  await server.connect(transport);
  // MCP server owns stdout — all other output must go to stderr
  console.error("PCR: MCP server running on stdio");
}

File: cmd/mcp.go
package cmd

import (
	"context"
	"fmt"
	"os"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"github.com/spf13/cobra"

	"github.com/pcr-developers/cli/internal/auth"
	"github.com/pcr-developers/cli/internal/supabase"
)

var mcpCmd = &cobra.Command{
	Use:   "mcp",
	Short: "Start the MCP server on stdio",
	RunE: func(cmd *cobra.Command, args []string) error {
		a := auth.Load()
		userID := ""
		token := ""
		if a != nil {
			userID = a.UserID
			token = a.Token
		}

		captureCount := 0

		s := server.NewMCPServer("pcr-dev", Version)

		// ── pcr_log_prompt ────────────────────────────────────────────────────
		s.AddTool(
			mcp.NewTool("pcr_log_prompt",
				mcp.WithDescription("Log a prompt and its AI response for later code review. Call this after completing a coding task."),
				mcp.WithString("prompt_text", mcp.Required(), mcp.Description("The user's prompt or instruction to the AI")),
				mcp.WithString("response_text", mcp.Description("The AI's response (summary or full text)")),
				mcp.WithString("session_id", mcp.Description("Session identifier to group related prompts")),
				mcp.WithString("project_name", mcp.Description("Name of the project being worked on")),
				mcp.WithString("branch_name", mcp.Description("Git branch name")),
				mcp.WithString("model", mcp.Description("AI model used (e.g., claude-sonnet-4-6)")),
				mcp.WithString("source", mcp.Description("AI tool name (e.g., claude-code, cursor, codex)")),
			),
			func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
				promptText, _ := req.Params.Arguments["prompt_text"].(string)
				responseText, _ := req.Params.Arguments["response_text"].(string)
				sessionID, _ := req.Params.Arguments["session_id"].(string)
				projectName, _ := req.Params.Arguments["project_name"].(string)
				branchName, _ := req.Params.Arguments["branch_name"].(string)
				model, _ := req.Params.Arguments["model"].(string)
				source, _ := req.Params.Arguments["source"].(string)

				if sessionID == "" {
					sessionID = fmt.Sprintf("mcp-%d", captureCount)
				}
				if source == "" {
					source = "unknown"
				}

				record := supabase.PromptRecord{
					SessionID:     sessionID,
					ProjectName:   projectName,
					BranchName:    branchName,
					PromptText:    promptText,
					ResponseText:  responseText,
					Model:         model,
					Source:        source,
					CaptureMethod: "mcp",
					UserID:        userID,
				}

				ok, err := supabase.UpsertPrompt(token, record)
				if ok {
					captureCount++
				}

				msg := fmt.Sprintf("PCR: Prompt logged (session total: %d)", captureCount)
				if err != nil || !ok {
					msg = "PCR: Failed to log prompt — check connection"
				}
				return mcp.NewToolResultText(msg), nil
			},
		)

		// ── pcr_log_session ───────────────────────────────────────────────────
		s.AddTool(
			mcp.NewTool("pcr_log_session",
				mcp.WithDescription("Log an entire coding session transcript. Use at the end of a session to capture all interactions at once."),
				mcp.WithString("session_id", mcp.Required(), mcp.Description("Unique session identifier")),
				mcp.WithString("project_name", mcp.Description("Project name")),
				mcp.WithString("branch_name", mcp.Description("Git branch")),
				mcp.WithString("source", mcp.Description("AI tool name")),
			),
			func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
				sessionID, _ := req.Params.Arguments["session_id"].(string)
				projectName, _ := req.Params.Arguments["project_name"].(string)
				branchName, _ := req.Params.Arguments["branch_name"].(string)
				source, _ := req.Params.Arguments["source"].(string)
				if source == "" {
					source = "unknown"
				}

				// messages is an array of {role, content, model?, timestamp?}
				messagesRaw, _ := req.Params.Arguments["messages"].([]any)
				var records []supabase.PromptRecord

				for i, msgRaw := range messagesRaw {
					msg, ok := msgRaw.(map[string]any)
					if !ok {
						continue
					}
					role, _ := msg["role"].(string)
					if role != "user" {
						continue
					}
					content, _ := msg["content"].(string)

					var responseText, model string
					if i+1 < len(messagesRaw) {
						if next, ok := messagesRaw[i+1].(map[string]any); ok {
							if nextRole, _ := next["role"].(string); nextRole == "assistant" {
								responseText, _ = next["content"].(string)
								model, _ = next["model"].(string)
							}
						}
					}
					if model == "" {
						model, _ = msg["model"].(string)
					}

					ts, _ := msg["timestamp"].(string)
					records = append(records, supabase.PromptRecord{
						SessionID:     sessionID,
						ProjectName:   projectName,
						BranchName:    branchName,
						PromptText:    content,
						ResponseText:  responseText,
						Model:         model,
						Source:        source,
						CaptureMethod: "mcp",
						CapturedAt:    ts,
						UserID:        userID,
					})
				}

				count, _ := supabase.UpsertPrompts(token, records)
				captureCount += count
				return mcp.NewToolResultText(
					fmt.Sprintf("PCR: Session logged — %d prompt(s) captured (session total: %d)", count, captureCount),
				), nil
			},
		)

		// ── pcr_status ────────────────────────────────────────────────────────
		s.AddTool(
			mcp.NewTool("pcr_status",
				mcp.WithDescription("Check how many prompts have been captured in this PCR session"),
			),
			func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
				user := userID
				if user == "" {
					user = "not logged in"
				}
				return mcp.NewToolResultText(
					fmt.Sprintf("PCR.dev status: %d prompt(s) captured this session. User: %s", captureCount, user),
				), nil
			},
		)

		fmt.Fprintln(os.Stderr, "PCR: MCP server running on stdio")
		return server.ServeStdio(s)
	},
}
